### DISCUSSION TITLE
A single client for multiple requests vs one client for each request

### DISCUSSION THREAD

**Discussion by fyse-nassar** [aws/aws-sdk-cpp#2606] [Q&A] [answered]
Can a  S3 client handle multiple requests in parallel? For instance, can a single client execute 10 getObjects calls simultaneously across different threads? Is this approach considered safe, or should each request be made using a unique client?

If the first scenario is correct, what are the limiting factors that might affect the client's performance? Are there any upper limits to the number of requests the client can effectively handle?

**Comment by jmklix** [comment, marked as answer]
@github-actions proposed-answer
>Can a S3 client handle multiple requests in parallel? 
>For instance, can a single client execute 10 getObjects calls simultaneously across different threads? 
>Is this approach considered safe, or should each request be made using a unique client?

Yes, the S3 client can handle multiple requests and it is considered safe. It is recommended to only ever use one client and to avoid using multiple clients.


>If the first scenario is correct, what are the limiting factors that might affect the client's performance? 
>Are there any upper limits to the number of requests the client can effectively handle?

The only limits is the network throughput and cpu. You should look into using the `PooledThreadExecutor` that will scale with the number of cores that you have. 

Please let me know if you have any other questions about this sdk.

**Reply by fyse-nassar** [reply]
Thanks for the reply. A few followup question.

> The only limits is the network throughput and cpu. You should look into using the PooledThreadExecutor that will scale with the number of cores that you have.

If I understand correctly, the `PooledThreadExecutor` is basically a pool of threads that are shared across multiple clients, unlike the default behaviour where each client spawns a thread and then detaches it.
So in case of a single client, how does the `PooledThreadExecutor` benefit?  Also, is `maxConnections` a parameter that affects the throughput?

**Reply by bhavya2109sharma** [reply]
- For single client the PooledThreadEcexutor will be beneficial as instead of creating new thread for each request, client can reuse threads from the pool, which will reduces the overhead of thread creation.

- maxConnections  parameter will affect the throughput as when making parallel requests with a single client, each request will acquire a connection from the connection pool. If the number of concurrent requests exceeds the maxConnections limit, additional requests will be queued until a connection becomes available.

- If you set maxConnections to a lower value, it can limit the throughput, as less requests can be processed concurrently.Although, setting it too high can also have negative consequences, as it can exhaust system resources.

- JFYI PooledThreadExecutor is primarily responsible for managing the threads used for executing requests, while the connection pool which controlled by maxConnections manages the HTTP connections used for communicating with the S3 service.
